File: src/main/java/modelLocal/Quiz.java
/*package modelLocal;

import com.iaroslaveremeev.quiz.repositories.QuestionRepository;
import com.iaroslaveremeev.quiz.util.Encrypt;
import util.Difficulty;

import java.util.List;
import java.util.Objects;

public class Quiz {

    private int numberOfQuestions;
    private Category category;
    private Difficulty difficulty;
    private List<Question> questions;

    public Quiz() {
    }

    public Quiz(int numberOfQuestions, Category category, Difficulty difficulty, List<Question> questions) {
        this.numberOfQuestions = numberOfQuestions;
        this.category = category;
        this.difficulty = difficulty;
        this.questions = questions;
    }

    public int getNumberOfQuestions() {
        return numberOfQuestions;
    }

    public void setNumberOfQuestions(int numberOfQuestions) {
        this.numberOfQuestions = numberOfQuestions;
    }

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public Difficulty getDifficulty() {
        return difficulty;
    }

    public void setDifficulty(Difficulty difficulty) {
        this.difficulty = difficulty;
    }

    public List<Question> getQuestions() {
        return questions;
    }

    public void setQuestions(List<Question> questions) {
        this.questions = questions;
    }

    *//*public void downloadQuestions() throws IOException {
        QuestionRepository questionRepository = new QuestionRepository();
        this.questions = questionRepository.downloadQuestions(this);
        for (Question question : questions) {
            String unescapedQuestion = StringEscapeUtils.unescapeHtml4(question.getQuestion());
            question.setQuestion(unescapedQuestion);
            String unescapedCorrectAnswer = StringEscapeUtils.unescapeHtml4(question.getCorrect_answer());
            question.setCorrect_answer(unescapedCorrectAnswer);
            String[] unescapedIncorrectAnswers = new String[question.getIncorrect_answers().length];
            for (int j = 0; j < question.getIncorrect_answers().length; j++) {
                unescapedIncorrectAnswers[j] = StringEscapeUtils.unescapeHtml4(question.getIncorrect_answers()[j]);
            }
            question.setIncorrect_answers(unescapedIncorrectAnswers);
        }
    }*//*

    // Encrypt questions from the quiz
    *//*public void encryptQuestions(int shift){
        ArrayList<Question> encryptedQuestions = new ArrayList<>();
        for (int i = 0; i < questions.size(); i++) {
            Question encryptedQuestion = Encrypt.encryptQuestion(questions.get(i), shift);
            encryptedQuestions.add(encryptedQuestion);
        }
        this.questions = encryptedQuestions;
    }*//*

    // Decrypt questions from the quiz
    *//*public void decryptQuestions(int shift){
        ArrayList<Question> decryptedQuestions = new ArrayList<>();
        for (int i = 0; i < questions.size(); i++) {
            Question decryptedQuestion = Encrypt.decryptQuestion(questions.get(i), shift);
            decryptedQuestions.add(decryptedQuestion);
        }
        this.questions = decryptedQuestions;
    }*//*

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Quiz quiz = (Quiz) o;
        return numberOfQuestions == quiz.numberOfQuestions && Objects.equals(category, quiz.category) && difficulty == quiz.difficulty && Objects.equals(questions, quiz.questions);
    }

    @Override
    public int hashCode() {
        return Objects.hash(numberOfQuestions, category, difficulty, questions);
    }

    @Override
    public String toString() {
        return "Quiz{" +
                "numberOfQuestions=" + numberOfQuestions +
                ", category=" + category +
                ", difficulty=" + difficulty +
                ", questions=" + questions +
                '}';
    }
}*/

File: src/main/webapp/scripts/model/quiz.js
import {Question} from './question.js';
import {Difficulty} from './difficulty.js';
import {QuestionRepository} from '../repositories/questionRepository.js';
import {Encrypt} from '../util/encrypt.js';
import he from 'he';
import {Category} from './category.js';

class Quiz {
    constructor(numberOfQuestions, category, difficulty, questions) {
        this.numberOfQuestions = numberOfQuestions;
        this.category = new Category(category);
        this.difficulty = difficulty;
        this.questions = questions;
    }

    setNumberOfQuestions(numberOfQuestions) {
        this.numberOfQuestions = numberOfQuestions;
    }

    getNumberOfQuestions() {
        return this.numberOfQuestions;
    }

    setCategory(category) {
        this.category = new Category(category);
    }

    getCategory() {
        return this.category;
    }

    setDifficulty(difficulty) {
        this.difficulty = difficulty;
    }

    getDifficulty() {
        return this.difficulty;
    }

    setQuestions(questions) {
        this.questions = questions;
    }

    getQuestions() {
        return this.questions;
    }

    async downloadQuestions() {
        const questionRepository = new QuestionRepository();
        this.questions = await questionRepository.downloadQuestions(this);
        for (let question of this.questions) {
            const unescapedQuestion = he.decode(question.getQuestion());
            question.setQuestion(unescapedQuestion);
            const unescapedCorrectAnswer = he.decode(question.getCorrect_answer());
            question.setCorrect_answer(unescapedCorrectAnswer);
            const unescapedIncorrectAnswers = new Array(question.getIncorrect_answers().length);
            for (let j = 0; j < question.getIncorrect_answers().length; j++) {
                unescapedIncorrectAnswers[j] = he.decode(question.getIncorrect_answers()[j]);
            }
            question.setIncorrect_answers(unescapedIncorrectAnswers);
        }
    }

    encryptQuestions(shift) {
        const encryptedQuestions = [];
        for (let i = 0; i < this.questions.length; i++) {
            const encryptedQuestion = Encrypt.encryptQuestion(this.questions[i], shift);
            encryptedQuestions.push(encryptedQuestion);
        }
        this.questions = encryptedQuestions;
    }

    decryptQuestions(shift) {
        const decryptedQuestions = [];
        for (let i = 0; i < this.questions.length; i++) {
            const decryptedQuestion = Encrypt.decryptQuestion(this.questions[i], shift);
            decryptedQuestions.push(decryptedQuestion);
        }
        this.questions = decryptedQuestions;
    }

    toString() {
        return `Quiz{ numberOfQuestions=${this.numberOfQuestions}, 
        category=${this.category.getName()}, difficulty=${this.difficulty}, 
        questions=${JSON.stringify(this.questions)} }`;
    }

}

export { Quiz };
